File: src/digraphx/parametric.py
from abc import abstractmethod
from fractions import Fraction
from typing import Generic, Mapping, MutableMapping, Tuple, TypeVar

from .neg_cycle import Cycle, Domain, Edge, NegCycleFinder, Node

Ratio = TypeVar("Ratio", Fraction, float)


class ParametricAPI(Generic[Node, Edge, Ratio]):
    @abstractmethod
    def distance(self, ratio: Ratio, edge: Edge) -> Ratio:
        """
        The `distance` function calculates the distance between a given ratio and edge.

        :param ratio: The `ratio` parameter is of type `Ratio`. It represents a ratio or proportion
        :type ratio: Ratio
        :param edge: The `edge` parameter represents an edge in a graph. It is of type `Edge`
        :type edge: Edge
        """

    @abstractmethod
    def zero_cancel(self, cycle: Cycle) -> Ratio:
        """
        The `zero_cancel` function takes a `Cycle` object as input and returns a `Ratio` object.

        :param cycle: The `cycle` parameter is of type `Cycle`.
        :type cycle: Cycle
        """


class MaxParametricSolver(Generic[Node, Edge, Ratio]):
    """Maximum Parametric Solver

    This class solves the following parametric network problem:

        max  r
        s.t. dist[v] - dist[u] <= distrance(e, r)
             forall e(u, v) in G(V, E)

    A parametric network problem refers to a type of optimization problem that
    involves finding the optimal solution to a network flow problem as a function
    of one single parameter.
    """

    def __init__(
        self,
        digraph: Mapping[Node, Mapping[Node, Edge]],
        omega: ParametricAPI[Node, Edge, Ratio],
    ) -> None:
        """
        The `__init__` function initializes an object with a graph and an omega parameter.

        :param digraph: digraph is a mapping of nodes to a mapping of nodes to edges. It represents a graph
            where each node is connected to other nodes through edges. The edges are represented by the
            mapping of nodes to edges

        :type digraph: Mapping[Node, Mapping[Node, Edge]]

        :param omega: The `omega` parameter is an instance of the `ParametricAPI` class. It represents
            some kind of parametric API that takes three type parameters: `Node`, `Edge`, and `Ratio`

        :type omega: ParametricAPI[Node, Edge, Ratio]
        """
        # self.ncf = NegCycleFinder(digraph)
        self.digraph = digraph
        self.omega: ParametricAPI[Node, Edge, Ratio] = omega

    def run(
        self, dist: MutableMapping[Node, Domain], ratio: Ratio
    ) -> Tuple[Ratio, Cycle]:
        """
        The `run` function takes in a distance mapping and a ratio, and iteratively finds the minimum
        ratio and corresponding cycle until the minimum ratio is greater than or equal to the input
        ratio.

        :param dist: The `dist` parameter is a mutable mapping where the keys are `Node` objects and the
            values are `Domain` objects. It represents the distance between nodes in a graph

        :type dist: MutableMapping[Node, Domain]

        :param ratio: The `ratio` parameter is a value that represents a ratio or proportion. It is used
            as a threshold or target value in the algorithm

        :type ratio: Ratio

        :return: The function `run` returns a tuple containing the updated ratio (`ratio`) and the cycle (`cycle`).
        """
        D = type(next(iter(dist.values())))

        def get_weight(e: Edge) -> Domain:
            return D(self.omega.distance(ratio, e))

        r_min = ratio
        c_min = []
        cycle = []

        ncf: NegCycleFinder[Node, Edge, Domain] = NegCycleFinder(self.digraph)

        while True:
            for ci in ncf.howard(dist, get_weight):
                ri = self.omega.zero_cancel(ci)
                if r_min > ri:
                    r_min = ri
                    c_min = ci
            if r_min >= ratio:
                break

            cycle = c_min
            ratio = r_min
        return ratio, cycle

File: rust_ai/src/parametric.rs
//! Parametric Network Solver
//!
//! This module defines a system for solving parametric network problems, which are a
//! type of optimization problem in graph theory. The main purpose of this code is
//! to find the maximum ratio that satisfies certain conditions in a graph, where
//! the distances between nodes depend on this ratio.

use std::collections::HashMap;

use crate::neg_cycle::NegCycleFinder;
use crate::types::{Cycle, Domain, Edge, Node, RatioType};

/// Parametric API trait for distance calculation and cycle analysis.
///
/// This trait defines an interface for how distances should be calculated and how
/// to find the ratio that makes a cycle's total distance zero.
pub trait ParametricAPI<N, E, R>
where
    N: Node,
    E: Edge,
    R: RatioType,
{
    /// Calculate the parametric distance for an edge given the current ratio.
    fn distance(&self, ratio: R, edge: &E) -> R;

    /// Calculate the actual ratio for a given cycle.
    ///
    /// This function calculates the ratio that would make the total distance of the cycle zero.
    fn zero_cancel(&self, cycle: &Cycle<E>) -> R;
}

/// Maximum Parametric Solver
///
/// This struct solves the following parametric network problem:
///
/// |    max  r
/// |    s.t. dist[v] - dist[u] <= distance(e, r)
/// |         forall e(u, v) in G(V, E)
///
/// A parametric network problem refers to a type of optimization problem that
/// involves finding the optimal solution to a network flow problem as a function
/// of one single parameter.
pub struct MaxParametricSolver<N, E, R, API>
where
    N: Node,
    E: Edge,
    R: RatioType,
    API: ParametricAPI<N, E, R>,
{
    /// The graph structure where nodes map to neighbors and edges
    digraph: HashMap<N, HashMap<N, E>>,
    /// Parametric API for distance calculation and cycle analysis
    omega: API,
    /// Marker for unused type parameter R
    _marker: std::marker::PhantomData<R>,
}

impl<N, E, R, API> MaxParametricSolver<N, E, R, API>
where
    N: Node,
    E: Edge,
    R: RatioType,
    API: ParametricAPI<N, E, R>,
{
    /// Initialize the solver with a graph and a parametric API.
    ///
    /// # Arguments
    ///
    /// * `digraph` - A mapping of nodes to a mapping of nodes to edges.
    /// * `omega` - A parametric API instance that provides methods for distance
    ///   calculation and cycle analysis.
    pub fn new(digraph: HashMap<N, HashMap<N, E>>, omega: API) -> Self {
        Self {
            digraph,
            omega,
            _marker: std::marker::PhantomData,
        }
    }

    /// Run the parametric network solver algorithm.
    ///
    /// The algorithm works by:
    /// 1. Starting with an initial ratio and distance estimates
    /// 2. Using a negative cycle finder to find cycles in the graph
    /// 3. For each negative cycle found, calculating a new ratio
    /// 4. Updating the best ratio if a better one is found
    /// 5. Repeating until no better ratio can be found
    ///
    /// # Arguments
    ///
    /// * `dist` - Initial distance labels for nodes
    /// * `ratio` - Initial ratio value to start the search
    ///
    /// # Returns
    ///
    /// A tuple containing the optimal ratio and the cycle that achieves it.
    pub fn run(&self, mut dist: HashMap<N, R>, ratio: R) -> (R, Cycle<E>)
    where
        R: Domain + Clone,
    {
        // Initialize minimum ratio and cycle
        let mut r_min = ratio.clone();
        let mut c_min = Vec::new();
        let mut cycle = Vec::new();
        let mut current_ratio = ratio.clone();

        // Create a negative cycle finder instance with the graph
        let mut ncf: NegCycleFinder<N, E, R> = NegCycleFinder::new(self.digraph.clone());

        // Main algorithm loop
        loop {
            // Define a weight function that calculates distance based on current ratio
            let get_weight = |e: &E| self.omega.distance(current_ratio.clone(), e);

            // Find all negative cycles in the graph
            let cycles = ncf.howard(&mut dist, &get_weight);
            for ci in cycles {
                // Calculate the ratio that would make this cycle's total distance zero
                let ri = self.omega.zero_cancel(&ci);
                
                // Update minimum ratio if a smaller one is found
                if r_min > ri {
                    r_min = ri;
                    c_min = ci;
                }
            }

            // Termination condition: no better ratio found
            if r_min >= current_ratio {
                break;
            }

            // Update cycle and ratio for next iteration
            cycle = c_min.clone();
            current_ratio = r_min.clone();
        }

        (current_ratio, cycle)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use num_rational::Ratio;
    use std::collections::HashMap;

    #[test]
    fn test_max_parametric_solver_new() {
        // Simple test API implementation
        struct TestAPI;

        impl ParametricAPI<&'static str, i32, Ratio<i64>> for TestAPI {
            fn distance(&self, ratio: Ratio<i64>, _edge: &i32) -> Ratio<i64> {
                // Simple implementation for testing
                ratio.clone()
            }

            fn zero_cancel(&self, _cycle: &Vec<i32>) -> Ratio<i64> {
                // Simple implementation for testing
                Ratio::new(1, 2)
            }
        }

        let _digraph: HashMap<&str, HashMap<&str, i32>> = HashMap::new();
        let api = TestAPI;
        let _solver = MaxParametricSolver::new(_digraph, api);
        
        // Just verify it compiles
        assert!(true);
    }
}